records/2024-02-06.AWS-vpn.md

```diff
@@ -0,0 +1,93 @@
+
+#### Pros
+
+- Can be managed in Terraform
+- Integrates directly with AWS CDS Single-Sign-On
+- VPN Client Is Cross Platform
+- Managed Service Requires Little Administration
+- The VPN endpoint protection is done by AWS
+
+#### Cons
+
+- Comparitively Expensive
+- Cloud specific means that we could not directly migrate this to another Provider
+
+#### Cost
+
+AWS charges per hour based on the number of subnet associations. The GC Notify Infrastructure would require associations with three private subnets. This results in a monthly charge of approximately $200 per month per AWS account. Given that GC Notify has 3 accounts (Dev/Staging/Prod), that results in $600 per month, or approximately $7200 per year for all three environments.
+
+#### Administration Overhead
+
+There is minimal administration required once it has been deployed. It is a fully managed service provided by AWS.
+
+### Self Hosted OpenVPN on Kubernetes
+
+It is possible to self-host an installation of [OpenVPN](https://openvpn.net/). OpenVPN is an industry standard, open source VPN solution that uses SSL communication. 
+
+It is possible to host this as a docker image in Kubernetes, and expose the endpoint. This would be roughly equivalent in network communication to how we currently do jump boxes. The helm chart is currently in a deprecated state, but the underlying docker image is still alive and well. It would be relatively simple to fork the deprecated helm chart and maintain our own copy.
+
+#### Pros
+- Since this would use existing resources, the cost would be very low, if not $0
+- Kubernetes would manage the container life-cycle and autoscaling
+
+#### Cons
+- There would be additional work for the core team to do, that would include patching
+- If the Kubernetes cluster experiences problems, there would be no way to connect to the VPN
+- There may be some additional network routing complications
+- Not a self managed solution
+- Does not directly integrate with CDS AWS SSO (Could be configured potentially with additional work)
+
+#### Cost
+
+This would be virtually free in terms of cloud spending since our Kubernetes cluster is constantly under utilized. It would be more expensive in salary time due to the maintenance overhead.
+
+### Self Hosted OpenVPN on an EC2
+
+Instead of containerizing and using OpenVPN on Kubernetes, we could create and deploy an EC2 instance and install OpenVPN on that.
+
+#### Pros
+- Solves the problem of being unable to connect if the Kubernetes cluster is down
+
+#### Cons
+- Autoscaling would not be configured (although we have so few users this may not be an issue)
+- Additional compute costs
+- Not a self managed solution
+- Would require an additional set of automation tools to facilitate the installation and configuration of OpenVPN (I.e Ansible)
+- Does not directly integrate with CDS AWS SSO (Could be configured potentially with additional work)
```

Comment:
Another thing to note this would go against the CDS Policy against IaaS and we would technically be required to implement SCED.

https://github.com/cds-snc/site-reliability-engineering/blob/main/documents/iaas_policy.md

Reply:
It would also potentially require us to perform  a new ATO as our existing is based on the usage of managed services and AFAIK doesn't handle hardening/patching VMs.

Reply:
This is major, definitely worth to list.

Reply:
Thanks Cal I added your suggestions.